### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive on evolving type

### ISSUE THREAD

**Issue by OpportunityLiu** [typescript-eslint/typescript-eslint#10334]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.tsx&code=LAKAJgpgxgNghgJwgAgGYFcB2UAuBLAe0zQIIAoBKALmU3QFsAjCBZAH2S0lT0wjFCRYiFBmz4iyRoko06TFoOjwkaLLkLFpALzKY5DZgmrIAbgTxgA3KFBiNk%2BnF4BGSsgDeoZMhgQcyAAeNiA%2BgcgAvFIyFCFhkSTksd7RuoEAhMkgAL62IPYSxE68AEzuXqG%2B-kFxQQnSCJS14VGopE0pOmSBWbkgQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```TypeScript
declare function foo(): number | undefined
declare function bar(): number
declare function baz(n: number): void;

function main1() {
  let x;
  x = bar();
  x = foo();
  baz(x!);
}

function main2() {
  let x;
  x = bar();
  x = foo();
  baz(x);
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

The linter should not treat `baz(x!)` as an unnecessary type assertion

### Actual Result

`@typescript-eslint/no-unnecessary-type-assertion` rule warns about the assertion baz(x!) as unnecessary, suggesting it could be removed. Removing the assertion causes TypeScript to raise an error.

**Comment by OpportunityLiu** [comment]
Maybe ralated: #7652 #8721 #10257

**Comment by bradzacher** [comment]
This is a case of strange behaviour from TS.
The type of your variable is a special kind of `any` that evolves and changes as you assign and use the variable.

[here is a playground with some annotations to show how things are weird](https://typescript-eslint.io/play/#ts=5.6.3&showAST=types&fileType=.tsx&code=LAKAJgpgxgNghgJwgAgGYFcB2UAuBLAe0zQIIAoBKALmU3QFsAjCBZAH2S0lT0wjFCRYiFBmz4iyRoko06TFoOjwkaLLkLFpALzKY5DZgmrIAbgTxgA3KFBiNk%2BnF4BGSsgDeoZMhgQcyAAeNiA%2BgcgAvFIyFCFhkSTksaAA9CkAegD83r7%2ByACeLgnBqSk%2BPlk5fgH5AEzFAIQhaeXIlaHRuoENySAt5e0AvrYg9hLETry17l4d1UFxQQnSCJSL4VGopGs5OmSBvf0V2SDDIEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false)

The important bit is that if you inspect the type of the identifier, `x` with and without a non-null assertion you'll see that TS itself reports weird types. When there is not a non-null assertion operator then the type of `x` is `number | undefined`, but when there is a non-null assertion operator then the type of `x` is just `number`.

To clarify -- it's specifically the identifier `x` that changes type, not the result of the non-null assertion operator!
This is why the rule reports - because we see a non-nullish type and thus based on the types the operator is unnecessary.

This might be a bug in TS, TBH. It certainly seems like very strange behaviour from TS.

**Comment by kirkwaiblinger** [comment]
I agree with @bradzacher that this looks like a TS bug. I would add, from playing with different TS versions in the playground, it appears that this behavior was introduced with 5.1. 

Seems like the next step here would be to get input from TS team to see if an unintentional regression occurred that can be fixed upstream.

**Comment by bradzacher** [comment]
There's another weird bit of behaviour I just noticed - when `noExplicitAny` is `false`, then TS never evolves the `any` and it's always `any`!!! [playground](https://www.typescriptlang.org/play/?strict=false&noImplicitAny=false#code/FAGwpgLgBAHg3MYMoF4oEYFIHQQPYByeEAojAJYDOEAFAJQJA)

**Comment by kirkwaiblinger** [comment]
@bradzacher do you mean `noImplicitAny`? This is expected https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html#improved-any-inference

**Comment by bradzacher** [comment]
Filed https://github.com/microsoft/TypeScript/issues/60514
Will close this for now.
We can reopen this if the result of that issue is working as intended and we can figure out a path forward.
